### DISCUSSION TITLE
Catch Up / Run All use case

### DISCUSSION THREAD

**Discussion by ian-kinner** [jhuckaby/Cronicle#685] [Q&A]
I have a fairly large schedule (~500 events) where most of them need to run - meaning if a job fails, it is either set to retry or is manually restarted. A few will be fine if they miss execution and run again later. Jobs currently run 24/7/365. I am trying to block off a couple hours to establish a maintenance window for my DB that all the jobs use. I'm thinking I could just disable the scheduler during this window to stop new jobs from starting, kill or wait out running jobs, and when the maintenance is done, turn the scheduler back on. If I understand correctly, jobs with the "Catch Up" attribute set will execute if they were supposed to run while I had the scheduler disabled. I have a few questions related to this:

1. Is my understanding of the catch up feature correct, or does it only kick off jobs if the server is down instead of the scheduler disabled?
2. I have ~500 events, and 0 of them currently have catch up enabled and this isn't an API-adjustable attribute. If I do an export and string replace "catch_up":0 with "catch_up":1, and then import that, will that update all the events, or append them to the scheduler as duplicates? Since the export contains the ID, I'm hoping/thinking it will update.
3. What are the ramifications of doing such a mass export/import? Will it affect running jobs?

Thanks in advance!

**Comment by jhuckaby** [comment]
Hello!

Yes, your understanding of the catch-up feature is correct.  If you disable the scheduler, then re-enable it later, it will run all the missed jobs that should have ran during the time when the scheduler was disabled.

> I have ~500 events, and 0 of them currently have catch up enabled and this isn't an API-adjustable attribute.

Yes it is!  You can pass `"catch_up":1` to the [update_event](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#update_event) API. 

> If I do an export and string replace "catch_up":0 with "catch_up":1, and then import that, will that update all the events, or append them to the scheduler as duplicates? Since the export contains the ID, I'm hoping/thinking it will update.

An import will replace / clobber all the records it imports, so your method may work, however...

> What are the ramifications of doing such a mass export/import? Will it affect running jobs?

You cannot perform an import while Cronicle is running.  Doing so would corrupt the entire DB.  It absolutely must be all stopped, and all jobs stopped, to do an import.

I highly recommend you script the [update_event](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#update_event) API instead.  First call [get_schedule](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#get_schedule) to get all your event IDs, then call [update_event](https://github.com/jhuckaby/Cronicle/blob/master/docs/APIReference.md#update_event) API for each one in sequence.  Example update call that enables catch-up:

```json
{
	"id": "3c182051",
	"catch_up": 1
}
```

All you need is the event `id`, and `catch_up`.  You can omit all other fields, as it performs a "sparse" update.

Good luck!
